### PR TITLE
Sn shadow assassin

### PR DESCRIPTION
**Pull request by Kert**

### CODE REVIEW

File: src/cpp/shared/combat/army.cpp [hunk 1]
```diff
@@ -1338,9 +1333,21 @@ void army::PowEffect(int animIdx, int a3, int a4, int a5) {
         } else {
           creature->field_3 = ANIMATION_TYPE_WINCE;
           creature->field_4 = ANIMATION_TYPE_WINCE_RETURN;
+          if(creature->creatureIdx == CREATURE_CYBER_SHADOW_ASSASSIN) { // astral dodge animations
+            if (!(creature->creature.creature_flags & RETALIATED) && !this->animatingRangedAttack) {
+              int dodgeAnimLen = 7;
+              creature->frameInfo.animationLengths[ANIMATION_TYPE_WINCE] = dodgeAnimLen;
+              for (int p = 0; p < dodgeAnimLen; p++)
+                creature->frameInfo.animationFrameToImgIdx[ANIMATION_TYPE_WINCE][p] = 34 + p;
```
Comment (jkoppel) [inline src/cpp/shared/combat/army.cpp]: nit: Please include braces

File: src/cpp/shared/combat/army.cpp [hunk 2]
```diff
@@ -1338,9 +1333,21 @@ void army::PowEffect(int animIdx, int a3, int a4, int a5) {
         } else {
           creature->field_3 = ANIMATION_TYPE_WINCE;
           creature->field_4 = ANIMATION_TYPE_WINCE_RETURN;
+          if(creature->creatureIdx == CREATURE_CYBER_SHADOW_ASSASSIN) { // astral dodge animations
+            if (!(creature->creature.creature_flags & RETALIATED) && !this->animatingRangedAttack) {
```
Comment (jkoppel) [inline src/cpp/shared/combat/army.cpp]: Reusing the RETALIATED flag? Clever, except that it doesn't quite work. This should also trigger for ranged attacks.

Comment (jkoppel) [inline src/cpp/shared/combat/army.cpp]: Please separate out this  animation logic.

File: src/cpp/shared/combat/army.cpp [hunk 3]
```diff
@@ -1623,4 +1634,146 @@ void army::MoveAttack(int targHex, int x) {
...
+    }
+  }
+}
+
+void army::InitClean() {
+  for (int i = 0; i < 7; ++i)
+    this->combatSounds[i] = 0;
+  this->lifespan = -1;
+  this->numActiveEffects = 0;
+  for(int i = 0; i < 19; i++) {
```
Comment (jkoppel) [inline src/cpp/shared/combat/army.cpp]: 19 is a magic number

File: src/cpp/shared/combat/army.cpp [hunk 4]
```diff
@@ -1623,4 +1634,146 @@ void army::MoveAttack(int targHex, int x) {
     CreatureHasAttribute(this->creatureIdx, STRIKE_AND_RETURN)) {
     MoveTo(startHex);
   }
+}
+
+void army::DecrementSpellRounds() {
+  for (int effect = 0; effect < 19; ++effect) {
```
Comment (jkoppel) [inline src/cpp/shared/combat/army.cpp]: Magic number

File: src/cpp/shared/combat/army.cpp (hunk 2 above)
Comment (Kert) [inline src/cpp/shared/combat/army.cpp]: So should the astral dodge be performed after ranged attacks or not? You said that maybe it's ok like this in skype

File: src/cpp/shared/scripting/consts.cpp [hunk 5]
```diff
@@ -320,6 +320,7 @@ void set_creature_consts(lua_State* L) {
   lua_setconst(L, "CREATURE_BLACK_KNIGHT", CREATURE_BLACK_KNIGHT);
   lua_setconst(L, "CREATURE_CATOBLEBA", CREATURE_CATOBLEBA);
   lua_setconst(L, "CREATURE_TREANT", CREATURE_TREANT);
+  lua_setconst(L, "CREATURE_CYBER_SHADOW_ASSASSIN", CREATURE_CYBER_SHADOW_ASSASSIN);
```
Comment (jkoppel) [inline src/cpp/shared/scripting/consts.cpp]: Nit: Swap the order of this and the Indigo panther so they'll be in the correct order

**Comment by jkoppel** [comment]
Well, the animation logic still isn't broken out, but it's going to be a bigger discussion how best to do that.

Move combatManager::CastSpell into combat_spells.cpp, and then you can merge.

**Pull request merged**